crates/accelerate/src/permutation.rs

```diff
@@ -77,6 +77,65 @@ fn get_ordered_swap(pattern: &ArrayView1<i64>) -> Vec<(i64, i64)> {
     swaps
 }
 
+fn pattern_to_cycles(pattern: &ArrayView1<i64>, invert_order: &bool) -> Vec<Vec<usize>> {
+    // vector keeping track of which elements in the permutation pattern have been visited
+    let mut explored: Vec<bool> = vec![false; pattern.len()];
+
+    // vector to store the cycles
+    let mut cycles: Vec<Vec<usize>> = Vec::new();
+
+    // cast the input pattern in terms of integers to usize, such that it can be used as index
+    // also invert the bit ordering if ``invert_order`` is true
+    let permutation: Array1<usize> = if *invert_order {
+        invert(pattern) // implies cast to usize
+    } else {
+        // TODO change to using Array1 insted of Vec
+        pattern.mapv(|x| x as usize)
+    };
+
+    for mut i in permutation.clone() {
+        let mut cycle: Vec<usize> = Vec::new();
+
+        // follow the cycle until we reached an entry we saw before
+        while !explored[i] {
+            cycle.push(i);
+            explored[i] = true;
+            i = permutation[i];
+        }
+        // cycles must have more than 1 element
+        if cycle.len() > 1 {
+            cycles.push(cycle);
+        }
+    }
```

Comment:
maybe this part could be done using iterators, or at least not cloning the permutation?

Reply:
I couldn't come up with an iterator solution since the we're changing the index during the iteration (which is why the compiler complains if `permutation` is not cloned), but I'm happy to discuss and see if we find a way 🙂

Reply:
You could avoid cloning the permutation, with the compiler complaints - guided solution ending up like this:
```
    for mut i in &permutation {
        let mut cycle: Vec<usize> = Vec::new();

        // follow the cycle until we reached an entry we saw before
        while !explored[*i] {
            cycle.push(*i);
            explored[*i] = true;
            i = &permutation[*i];
        }
        // cycles must have more than 1 element
        if cycle.len() > 1 {
            cycles.push(cycle);
        }
    }
```
But maybe there is an even cleaner solution.

Reply:
Or this (this compiles but I did not run it):
```
fn pattern_to_cycles(pattern: &ArrayView1<i64>, invert_order: &bool) -> Vec<Vec<usize>> {
    // vector keeping track of which elements in the permutation pattern have been visited
    let mut explored: Vec<bool> = vec![false; pattern.len()];

    // cast the input pattern in terms of integers to usize, such that it can be used as index
    // also invert the bit ordering if ``invert_order`` is true
    let permutation: Array1<usize> = if *invert_order {
        invert(pattern) // implies cast to usize
    } else {
        pattern.mapv(|x| x as usize)
    };

    permutation.iter().map(|pos| {
        let mut cycle: Vec<usize> = Vec::new();

        // follow the cycle until we reached an entry we saw before
        let mut i = *pos;
        while !explored[i] {
            cycle.push(i);
            explored[i] = true;
            i = permutation[i];
        }
        cycle
    }).filter(|c| c.len() > 1).collect()
}
```

Reply:
Lol, I don't think this is very efficient, but you can rewrite the inner while-loop as an iterator too, something along the following lines:
```

fn pattern_to_cycles(pattern: &ArrayView1<i64>, invert_order: &bool) -> Vec<Vec<usize>> {
    // vector keeping track of which elements in the permutation pattern have been visited
    let mut explored: Vec<bool> = vec![false; pattern.len()];

    // cast the input pattern in terms of integers to usize, such that it can be used as index
    // also invert the bit ordering if ``invert_order`` is true
    let permutation: Array1<usize> = if *invert_order {
        invert(pattern) // implies cast to usize
    } else {
        pattern.mapv(|x| x as usize)
    };

    permutation
        .iter()
        .map(|pos| {
            // follow the cycle until we reached an entry we saw before
            std::iter::successors(if !explored[*pos] { Some(*pos) } else { None }, |&i| {
                explored[i] = true;
                if !explored[permutation[i]] {
                    Some(permutation[i])
                } else {
                    None
                }
            })
            .collect::<Vec<usize>>()
        })
        .filter(|c| c.len() > 1)
        .collect()
}
```

Reply:
As discussed offline, iterators turned out not to speed up the code, so for a simpler code we'll stick to the current version. The iterator-only version is available here if anyone is interested: https://github.com/Cryoris/qiskit-terra/blob/92d91440a9c970410ebe337f8cf425bcb065c999/crates/accelerate/src/synthesis/permutation/utils.rs#L101 🙂